### PR TITLE
Add simple discovery demo

### LINKED ISSUE #108: Add basic examples
A small command-line .NET app should be sufficient.

### PR DESCRIPTION
**Pull request by fibann**
Resolves #108.

### CODE REVIEW

File: libs/Matchmaking/examples/DiscoveryDemo/Program.cs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+        Program(string username, IPAddress broadcastAddress)
+        {
+            _username = username;
+            _localAddrString = GetLocalIPAddress().ToString();
+            _chatServer = new TcpListener(IPAddress.Any, ChatPort);
+            ListenForConnectionsAsync();
+
+            // Initialize the discovery agent
+            _discoveryAgent = new PeerDiscoveryAgent(new UdpPeerDiscoveryTransport(broadcastAddress, DiscoveryPort));
```
[the remark this answers is not in the document]
Comment (fibann) [reply libs/Matchmaking/examples/DiscoveryDemo/Program.cs]: Yes, I'll try to compact the relevant code all in one place.

File: libs/Matchmaking/examples/DiscoveryDemo/Program.cs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+            // Subscribe to other participant resources.
+            _discoverySubscription = _discoveryAgent.Subscribe(ParticipantCategory);
+            _discoverySubscription.Updated += subscription =>
+            {
+                // When the active resources change, refresh the publishers list.
+                RefreshPublishers(_discoverySubscription.Resources);
+            };
+
+            // Initialize the publishers list.
+            RefreshPublishers(_discoverySubscription.Resources);
```
[the remark this answers is not in the document]
Comment (fibann) [reply libs/Matchmaking/examples/DiscoveryDemo/Program.cs]: See https://github.com/microsoft/MixedReality-Sharing/issues/39#issuecomment-524793096 - this was one of the original options. My worry was that calling the handler in `+=` was un-idiomatic and surprising. Let's discuss IRL before reopening the issue.

File: libs/Matchmaking/examples/DiscoveryDemo/Program.cs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        private const string ParticipantCategory = "Microsoft.MixedReality.Sharing.Matchmaking.DiscoveryDemo/participant";
+        private const string NameKey = "name";
+
+        private class Publisher
+        {
+            public string Name;
+            public TcpClient Client;
+        }
+
+        private List<TcpClient> _subscribers = new List<TcpClient>();
```
[the remark this answers is not in the document]
Comment (fibann) [reply libs/Matchmaking/examples/DiscoveryDemo/Program.cs]: The reason for the two sockets is that doing a handshake to figure out who has to be the client/the server is complicated 😄 Yes, it deserves an explanation. 

On second thought I suppose that I can have the lower IP address connect to the higher, that might require less text to explain but will need some way to attach the correct username to each incoming connection. Will give it a try.

[the remark this answers is not in the document]
Comment (stephenatwork) [reply libs/Matchmaking/examples/DiscoveryDemo/Program.cs]: I don't know if this needs to be fixed for the demo. I like that it's P2P in more ways than one.

File: libs/Matchmaking/examples/DiscoveryDemo/Program.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fibann) [reply libs/Matchmaking/examples/DiscoveryDemo/Program.cs]: Should be nicer now

File: libs/Matchmaking/examples/DiscoveryDemo/Program.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fibann) [reply libs/Matchmaking/examples/DiscoveryDemo/Program.cs]: In the end I decided to leave the two sockets and add some comments - handshakes are hard.

**Comment by fibann** [comment]
This PR cannot be updated anymore since the fork became broken after the main repo became public. So I'll merge this PR as it is and follow up with a new one.

**Pull request merged**